### PR TITLE
Added an explicit field in the function details for componenttype

### PR DESCRIPTION
**Pull request by srkukarni**
### Motivation
Sources/Sinks/Functions all use the same underlying representation of FunctionDetails to store their metadata. This is currently missing a componentType, which tells whether its a function/source/sink. Right now we kind of infer it with some heuristics but its better to explicitly mark it. This pr adds that field.

### Modifications

*Describe the modifications you've done.*

### Verifying this change

- [ ] Make sure that the change passes the CI checks.

*(Please pick either of the following options)*

This change is a trivial rework / code cleanup without any test coverage.

*(or)*

This change is already covered by existing tests, such as *(please describe tests)*.

*(or)*

This change added tests and can be verified as follows:

*(example:)*
  - *Added integration tests for end-to-end deployment with large payloads (10MB)*
  - *Extended integration test for recovery after broker failure*

### Does this pull request potentially affect one of the following parts:

*If `yes` was chosen, please highlight the changes*

  - Dependencies (does it add or upgrade a dependency): (yes / no)
  - The public API: (yes / no)
  - The schema: (yes / no / don't know)
  - The default values of configurations: (yes / no)
  - The wire protocol: (yes / no)
  - The rest endpoints: (yes / no)
  - The admin cli options: (yes / no)
  - Anything that affects deployment: (yes / no / don't know)

### Documentation

  - Does this pull request introduce a new feature? (yes / no)
  - If yes, how is the feature documented? (not applicable / docs / JavaDocs / not documented)
  - If a feature is not applicable for documentation, explain why?
  - If a feature is not documented yet in this PR, please create a followup issue for adding the documentation

### CODE REVIEW

**Comment by srkukarni** [comment]
run cpp tests

**Comment by srkukarni** [comment]
run java8 tests

**Comment by srkukarni** [comment]
run java8 tests

**Comment by srkukarni** [comment]
run java8 tests

**Comment by srkukarni** [comment]
run integration tests

**Comment by srkukarni** [comment]
run integration tests
run java8 tests

**Comment by srkukarni** [comment]
run integration tests
run java8 tests

File: pulsar-functions/worker/src/main/java/org/apache/pulsar/functions/worker/rest/api/ComponentImpl.java [hunk 1]
```diff
@@ -66,11 +66,7 @@
 import org.apache.pulsar.functions.proto.Function.SourceSpec;
 import org.apache.pulsar.functions.proto.InstanceCommunication;
 import org.apache.pulsar.functions.runtime.RuntimeSpawner;
-import org.apache.pulsar.functions.utils.ComponentType;
-import org.apache.pulsar.functions.utils.FunctionCommon;
-import org.apache.pulsar.functions.utils.FunctionConfigUtils;
-import org.apache.pulsar.functions.utils.SinkConfigUtils;
-import org.apache.pulsar.functions.utils.SourceConfigUtils;
+import org.apache.pulsar.functions.utils.*;
```
Comment (jerrypeng) [inline pulsar-functions/worker/src/main/java/org/apache/pulsar/functions/worker/rest/api/ComponentImpl.java]: Its not good practice to use wild cards imports

File: pulsar-functions/instance/src/main/java/org/apache/pulsar/functions/instance/InstanceUtils.java [hunk 2]
```diff
@@ -89,23 +85,26 @@
         }
     }
 
-    public ComponentType calculateSubjectType(Function.FunctionDetails functionDetails) {
+    public Function.FunctionDetails.ComponentType calculateSubjectType(Function.FunctionDetails functionDetails) {
```
Comment (jerrypeng) [inline pulsar-functions/instance/src/main/java/org/apache/pulsar/functions/instance/InstanceUtils.java]: can we add a unit test for this to make sure there isn't any issues with bc

File: pulsar-functions/worker/src/main/java/org/apache/pulsar/functions/worker/rest/api/ComponentImpl.java (hunk 1 above)
Comment (srkukarni) [reply pulsar-functions/worker/src/main/java/org/apache/pulsar/functions/worker/rest/api/ComponentImpl.java]: removed

File: pulsar-functions/instance/src/main/java/org/apache/pulsar/functions/instance/InstanceUtils.java (hunk 2 above)
Comment (srkukarni) [reply pulsar-functions/instance/src/main/java/org/apache/pulsar/functions/instance/InstanceUtils.java]: unittest added

**Comment by srkukarni** [comment]
run integration tests
run java8 tests

**Comment by srkukarni** [comment]
run java8 tests
run integration tests

**Comment by srkukarni** [comment]
run cpp tests

**Pull request merged**